chronos/streams/chunkstream.nim

```diff
@@ -0,0 +1,383 @@
+                     fut3: Future[C]): Future[void] =
+  ## Wait for completion of one future from list [`fut1`, `fut2`, `fut3`],
+  ## resulting future do not care about futures error, so you need to check
+  ## `fut1`, `fut2`, `fut3` for error manually.
+  var retFuture = newFuture[void]("chunked.oneOf()")
+  proc cb(data: pointer) {.gcsafe.} =
+    var index: int
+    if not retFuture.finished:
+      var fut = cast[FutureBase](data)
+      if cast[pointer](fut1) == data:
+        fut2.removeCallback(cb)
+        fut3.removeCallback(cb)
+      elif cast[pointer](fut2) == data:
+        fut1.removeCallback(cb)
+        fut3.removeCallback(cb)
+      elif cast[pointer](fut3) == data:
+        fut2.removeCallback(cb)
+        fut3.removeCallback(cb)
+      retFuture.complete()
+  fut1.callback = cb
+  fut2.callback = cb
+  fut3.callback = cb
+  return retFuture
+
+proc oneOf*[A, B](fut1: Future[A], fut2: Future[B]): Future[void] =
+  ## Wait for completion of `fut1` or `fut2`, resulting future do not care about
+  ## error, so you need to check `fut1` and `fut2` for error.
+  var retFuture = newFuture[void]("chunked.oneOf()")
+  proc cb(data: pointer) {.gcsafe.} =
+    var index: int
+    if not retFuture.finished:
+      var fut = cast[FutureBase](data)
+      if cast[pointer](fut1) == data:
+        fut2.removeCallback(cb)
+      elif cast[pointer](fut2) == data:
+        fut1.removeCallback(cb)
+      retFuture.complete()
+  fut1.callback = cb
+  fut2.callback = cb
+  return retFuture
+
+proc getChunkSize(buffer: openarray[byte]): uint64 =
+  # We using `uint64` representation, but allow only 2^32 chunk size,
+  # ChunkHeaderSize.
+  for i in 0..<min(len(buffer), ChunkHeaderSize):
+    let ch = buffer[i]
+    if char(ch) in {'0'..'9', 'a'..'f', 'A'..'F'}:
+      if ch >= byte('0') and ch <= byte('9'):
+        result = (result shl 4) or uint64(ch - byte('0'))
+      else:
+        result = (result shl 4) or uint64((ch and 0x0F) + 9)
+    else:
+      result = 0xFFFF_FFFF_FFFF_FFFF'u64
+      break
+
+proc setChunkSize(buffer: var openarray[byte], length: int64): int =
+  # Store length as chunk header size (hexadecimal value) with CRLF.
+  # Maximum stored value is ``0xFFFF_FFFF``.
+  # Buffer ``buffer`` length must be at least 10 octets.
+  doAssert(length <= int64(uint32.high))
+  var n = 0xF000_0000'i64
+  var i = 32
+  var c = 0
+  if length == 0:
+    buffer[0] = byte('0')
+    buffer[1] = byte(0x0D)
+    buffer[2] = byte(0x0A)
+    result = 3
+  else:
+    while n != 0:
+      var v = length and n
+      if v != 0 or c != 0:
+        let digit = byte((length and n) shr (i - 4))
+        var ch = digit + byte('0')
+        if ch > byte('9'):
+          ch = ch + 0x07'u8
+        buffer[c] = ch
+        inc(c)
+      n = n shr 4
+      i = i - 4
+    buffer[c] = byte(0x0D)
+    buffer[c + 1] = byte(0x0A)
+    result = c + 2
+
+proc chunkedReadLoop(stream: AsyncStreamReader) {.async.} =
+  var rstream = cast[ChunkedStreamReader](stream)
```

Comment:
This proc has a lot of boilerplate for a relatively simple format. Why can't we hide all the error conditions such "stream closed", "operation timed-out", etc behind a simpler "attempt-to-read()" API?

Reply:
@zah could you please explain more on `attempt-to-read` API?